### ISSUE TITLE
AFQMC converter failures

### ISSUE THREAD

**Issue by prckent** [QMCPACK/qmcpack#2851]
**Describe the bug**

Many of the AFQMC converter tests are failing in the nightlies, e.g. https://cdash.qmcpack.org/CDash/viewTest.php?onlyfailed&buildid=177579

There could be configuration issues here (which we should trap) or underlying bugs. Pyscf may be at fault for a few. Possibly the converter scripts need updating for h5py 3?

**To Reproduce**

Develop version, gcc 10 mkl complex and real. python 3.8.7, h5py 3.1.0...., all installed via spack. I can push the exact nightly script and spack hash etc. if needed, but they are essentially the canonical ones.

**Expected behavior**
Tests pass

**System:**

Our nightly test systems sulfur, nitrogen

**Additional context**

**Comment by prckent** [comment]
@Paul-St-Young Created as requested. Very interested if you can reproduce any of these. Suggestive of version dependencies/missed updates are errors such as https://cdash.qmcpack.org/CDash/testDetails.php?test=11508899&build=177579

**Comment by Paul-St-Young** [comment]
@prckent Thanks! I will try to fix these. Gotta learn spack first

**Comment by prckent** [comment]
Knowing whether they run on a random e.g. Ubuntu would also help. There shouldn't be anything spack specific here.

**Comment by prckent** [comment]
Note that h5py v3 is needed to support the latest HDF5 v1.12.0

**Comment by Paul-St-Young** [comment]
Yes, I was able reproduce many breakages on a fresh install of Fedora 33 (gcc 10.2.1, python 3.9.1, h5py 3.1.0 )

07 and 08 breakages are trivial and due to change of default mode of `h5py.File` from "r+" to "r". Solution is to specify "r+" mode `h5py.File('scf.chk', 'r+')`.

06 breakage is also trivial and due to removal of "fmt" keyword from `plt.savefig` in matplotlib 3.3. Solution is to remove this keyword.

I cannot reproduce problem with 01 using pyscf 1.7.5 from pip install. This appears to be a PySCF problem anyways.

05 is only real problem. I'll look into this tomorrow

**Comment by Paul-St-Young** [comment]
`SparseTensorIO.hpp::writeSparseTensor` under the flag `write_hdf` in `FactorizedSparseHamiltonian.cpp` is the source of the breakage for example 05.

I got "ERROR HDF5 write failure in hdf_archive::write dims".
My guess is an analogous default mode changing from "r+" to "r" when opening an HDF5 file.

Hopefully an easy fix.

**Comment by prckent** [comment]
Can you PR the easy fixes that you have already? Thanks.

**Comment by Paul-St-Young** [comment]
sorry I've been dragging my feet with the fix

To fix the HDF5 error: Enable both read and write on hdf_archive `dump` 
```
Wavefunction WavefunctionFactory::fromHDF5
-  if (!dump.open(filename, H5F_ACC_RDONLY))
+  if (!dump.open(filename, H5F_ACC_RDWR))
```

Not sure how this was working before. I guess `dump` allowed writing even though the file was opened with read-only mode?
@mmorale3 perhaps we a separate HDF5 file for checkpointing

Will open a PR now.
